recoil2png.1
.TH RECOIL2PNG 1 "Jan 12, 2021" "Retro Computer Image Library"
.SH NAME
recoil2png \- convert retro computer images to PNG
.SH SYNOPSIS
.B recoil2png
.B [\-o
.I OUTPUTFILE
.B ]
.I INPUTFILE...
.SH DESCRIPTION
.B recoil2png
converts pictures in native formats of vintage computers to the PNG format.
If multiple input files are specified, each one is converted separately.

.B recoil2png
recognizes over 400 formats native to the following computers:
Amiga, Amstrad CPC, Apple II, Atari 8-bit, Atari Portfolio, Atari ST/TT/Falcon,
BBC Micro, Commodore VIC-20, Commodore 16, Commodore 64, Commodore 128,
Electronika BK, FM Towns, Macintosh 128K, MSX, NEC PC-80/88/98, Oric,
Psion Series 3, SAM Coupe, Sharp X68000, Timex 2048, TRS-80,
TRS-80 Color Computer, ZX81 and ZX Spectrum.
See http://recoil.sourceforge.net/ for the full list of formats.

Input files must have the filename extension that matches the format.

Some pictures are composed of multiple files \- they must share their
filename and only differ by their extensions.  You should only pass
the filename of the "main" file.

.SH OPTIONS
.TP
\fB\-o\fR \fIFILE\fR, \fB\-\-output\fR=\fIFILE\fR
Set output PNG file name.
If you do not specify this option, the name of the input file is reused,
with the extension changed to "png".
.TP
\fB\-p\fR \fIFILE\fR, \fB\-\-palette\fR=\fIFILE\fR
Use the specified RGB palette for Atari 8-bit or C64 pictures.
Atari 8-bit palette files must be 768 bytes long with
an extension of "act" or "pal".
C64 palette files must be text files in the VICE Palette format
with the extension "vpl".
.TP
\fB\-h\fR, \fB\-\-help\fR
Display help message and exit.
.TP
\fB\-v\fR, \fB\-\-version\fR
Display version information and exit.
.SH COPYRIGHT
.B recoil2png
is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published
by the Free Software Foundation; either version 2 of the License,
or (at your option) any later version.


debian/changelog
recoil (5.1.1-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Fri, 30 Oct 2020 10:37:43 +0100

recoil (5.1.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Tue, 8 Sep 2020 14:15:21 +0200

recoil (5.0.1-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Wed, 18 Mar 2020 17:37:54 +0100

recoil (5.0.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Tue, 12 Nov 2019 09:35:17 +0100

recoil (4.3.2-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Sat, 6 Jul 2019 09:58:32 +0200

recoil (4.3.1-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Wed, 5 Dec 2018 21:08:30 +0100

recoil (4.3.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Sun, 10 Jun 2018 11:17:46 +0200

recoil (4.2.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Tue, 20 Feb 2018 16:06:42 +0100

recoil (4.1.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Wed, 18 Oct 2017 18:45:04 +0100

recoil (4.0.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Wed, 17 May 2017 13:21:57 +0200

recoil (3.5.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Thu, 1 Dec 2016 18:15:54 +0100

recoil (3.4.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Tue, 2 Feb 2016 09:27:52 +0100

recoil (3.3.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Fri, 7 Aug 2015 17:56:32 +0200

recoil (3.2.0-1) UNRELEASED; urgency=low

  * New release.

 -- Piotr Fusik <fox@scene.pl>  Wed, 28 Jan 2015 10:42:15 +0100

recoil (3.1.0-1) UNRELEASED; urgency=low

  * First deb packaging.

 -- Piotr Fusik <fox@scene.pl>  Wed, 25 Jun 2014 09:58:19 +0200


recoil2png.c
/*
 * recoil2png.c - command-line converter of retro computer pictures to the PNG format
 *
 * Copyright (C) 2009-2021  Piotr Fusik and Adrian Matoga
 *
 * This file is part of RECOIL (Retro Computer Image Library),
 * see http://recoil.sourceforge.net
 *
 * RECOIL is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * RECOIL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with RECOIL; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "recoil-stdio.h"
#include "pngsave.h"

static void print_help(void)
{
	printf(
		"Usage: recoil2png [OPTIONS] INPUTFILE...\n"
		"Options:\n"
		"-o FILE  --output=FILE   Set output file name\n"
		"         --pal           Emulate PAL video standard if applicable (default)\n"
		"         --ntsc          Emulate NTSC video standard if applicable\n"
		"-p FILE  --palette=FILE  Load Atari 8-bit or C64 palette\n"
		"-h       --help          Display this information\n"
		"-v       --version       Display version information\n"
	);
}

static int load_file(const char *filename, void *buffer, size_t buffer_len)
{
	FILE *fp = fopen(filename, "rb");
	if (fp == NULL) {
		fprintf(stderr, "recoil2png: cannot open %s\n", filename);
		return -1;
	}
	int len = fread(buffer, 1, buffer_len, fp);
	fclose(fp);
	return len;
}

static bool load_palette(RECOIL *recoil, const char *filename)
{
	uint8_t content[RECOIL_MAX_PLATFORM_PALETTE_CONTENT_LENGTH];
	int content_len = load_file(filename, content, sizeof(content));
	if (content_len < 0) {
		/* error already printed */
		return false;
	}
	if (!RECOIL_SetPlatformPalette(recoil, filename, content, content_len)) {
		fprintf(stderr, "recoil2png: %s: invalid palette file\n", filename);
		return false;
	}
	return true;
}

static bool process_file(RECOIL *recoil, const char *input_file, const char *output_file)
{
	static uint8_t content[RECOIL_MAX_CONTENT_LENGTH];
	int content_len = load_file(input_file, content, sizeof(content));
	if (content_len < 0) {
		/* error already printed */
		return false;
	}
	if (!RECOIL_Decode(recoil, input_file, content, content_len)) {
		fprintf(stderr, "recoil2png: %s: file decoding error\n", input_file);
		return false;
	}

	if (output_file == NULL) {
		static char output_default[FILENAME_MAX];
		int i;
		int dotp = 0;
		for (i = 0; input_file[i] != '\0' && i < FILENAME_MAX - 5; i++)
			if ((output_default[i] = input_file[i]) == '.')
				dotp = i;
		strcpy(output_default + (dotp == 0 ? i : dotp), ".png");
		output_file = output_default;
	}
	FILE *fp = fopen(output_file, "wb");
	if (fp == NULL || !RECOIL_SavePng(recoil, fp)) {
		fprintf(stderr, "recoil2png: cannot write %s\n", output_file);
		return false;
	}
	return true;
}

int main(int argc, char **argv)
{
	RECOIL *recoil = RECOILStdio_New();
	if (recoil == NULL) {
		fprintf(stderr, "recoil2png: out of memory\n");
		return 1;
	}
	const char *output_file = NULL;
	bool ok = true;
	bool no_input_files = true;
	for (int i = 1; i < argc; i++) {
		const char *arg = argv[i];
		if (arg[0] != '-') {
			ok &= process_file(recoil, arg, output_file);
			no_input_files = false;
			output_file = NULL;
		}
		else if (arg[1] == 'o' && arg[2] == '\0' && i + 1 < argc)
			output_file = argv[++i];
		else if (strncmp(arg, "--output=", 9) == 0)
			output_file = arg + 9;
		else if (strcmp(arg, "--pal") == 0)
			RECOIL_SetNtsc(recoil, false);
		else if (strcmp(arg, "--ntsc") == 0)
			RECOIL_SetNtsc(recoil, true);
		else if (arg[1] == 'p' && arg[2] == '\0' && i + 1 < argc) {
			if (!load_palette(recoil, argv[++i]))
				return 1;
		}
		else if (strncmp(arg, "--palette=", 10) == 0) {
			if (!load_palette(recoil, arg + 10))
				return 1;
		}
		else if ((arg[1] == 'h' && arg[2] == '\0')
			|| strcmp(arg, "--help") == 0) {
			print_help();
			no_input_files = false;
		}
		else if ((arg[1] == 'v' && arg[2] == '\0')
			|| strcmp(arg, "--version") == 0) {
			printf("recoil2png " RECOIL_VERSION "\n");
			no_input_files = false;
		}
		else {
			fprintf(stderr, "recoil2png: unknown option: %s\n", arg);
			return 1;
		}
	}
	if (no_input_files) {
		print_help();
		return 1;
	}
	return ok ? 0 : 1;
}
